File: programmers/kakao#2/java/KAKAO26.java
class KAKAO26 {
    public String solution(String play_time, String adv_time, String[] logs) {
        int playTime = mapToSecond(play_time);
        int advTime = mapToSecond(adv_time);
        long[] playCnt = new long[playTime + 1];
        
        for (String log : logs) {
            int start = mapToSecond(log.split("-")[0]);
            int end = mapToSecond(log.split("-")[1]);
            
            // 시작 시간과 끝나는 시간 마킹
            playCnt[start]++;
            playCnt[end]--;
        }
        
        // 누적 합
        for (int i = 1;i < playTime;i++) {
            playCnt[i] += playCnt[i - 1];
        }
        
        
        long currentCnt = 0L;
        for (int i = 0; i < advTime; i++) {
            currentCnt += playCnt[i];    
        }
        
        long maxCnt = currentCnt, maxIdx = 0;
        for (int right = advTime; right < playTime; right++) {
            currentCnt = currentCnt + playCnt[right] - playCnt[right - advTime];
            
            if (currentCnt > maxCnt) {
                maxCnt = currentCnt;
                maxIdx = right - advTime + 1;
            }
        }
        
        return mapToString(maxIdx);
    }
    
    private int mapToSecond(String time) {
        int hour = Integer.parseInt(time.split(":")[0]) * 60 * 60;
        int minuet = Integer.parseInt(time.split(":")[1]) * 60;
        int second = Integer.parseInt(time.split(":")[2]);
        
        return hour + minuet + second;
    }
    
    private String mapToString(long time) {
        long hour = time / (60 * 60);
        long minuet = (time / 60) % 60;
        long second = time % 60;
        
        return String.format("%02d:%02d:%02d", hour, minuet, second);
    }
}
/**
 * [카카오 2021 공채] 광고 삽입
 * 해결: X
 * 
 */

File: 0_4.programmers/kakao#2/python/KAKAO26.py
from collections import defaultdict

def solution(play_time, adv_time, logs):
    play_time = convert(play_time)
    adv_time = convert(adv_time)
    total_second = defaultdict(int)
    
    for log in logs:
        log_start, log_end = log.split("-")
        total_second[convert(log_start)] += 1
        total_second[convert(log_end)] -= 1
    
    for i in range(1, play_time):
        total_second[i] += total_second[i - 1]
        
    current_cnt = 0
    for i in range(1, adv_time):
        current_cnt += total_second[i]
        
    max_cnt = current_cnt
    max_idx = 0
    
    for i in range(adv_time, play_time):
        current_cnt = current_cnt + total_second[i] - total_second[i - adv_time]
        
        if current_cnt > max_cnt:
            max_cnt = current_cnt
            max_idx = i - adv_time + 1
            
    answer = '%02d:%02d:%02d' % (max_idx / 3600, (max_idx / 60) % 60, max_idx % 60)
    return answer

def convert(time):
    h, m, s  = map(int, time.split(":"))
    return h * 60 * 60 + m * 60 + s

'''
[카카오 2021 공채] 광고 삽입
해결: X
'''